add scaleMapper property to series & optimise performance

chrisprice opened the pull request:
Still to do -

* ~Add `scaleMapper` property to series~

Offers an alternative to #1492 which offers a route forward for #1491.

I debated for a long time whether `equals` & `scaleMapper` should be cascaded by `seriesWebglMulti` before realising it didn't exist. I'd come to the conclusion that it probably should be cascaded -

* `equals` - I think in most cases you'd want consistent behaviour across all series but if you didn't it would be possible to have a consistent check (e.g. `obj[Symbol.equals](other)`) and use `mapping` to mutate this value to have different behaviour per series
* `scaleMapper` - I can't see a use case for a different `scaleMapper` per series

cprice-scottlogic committed:
refactor: return "immutable" scales from scaleMapper

cprice-scottlogic committed:
feat: add scaleMapper property

cprice-scottlogic committed:
docs: add scaleMapper & equals docs

OliverForeman commented on packages/d3fc-series/README.md:
```diff
@@ -429,6 +429,20 @@ The SVG and canvas implementations do not support this property.
+
+If *scaleMapper* is specified, sets the function used to map first the xScale and then the yScale onto matched pairs of JavaScript and WebGL implementations. If *equals* is not specified, returns the current scale mapper which defaults to [webglScaleMapper](https://github.com/d3fc/d3fc/packages/d3fc-webgl#webglScaleMapper).
+
+The returned JavaScript scale is equlity checked using a reference comparison to determine whether the data needs to be rescaled. If the reference is the same as the previous render, the values associated with the scale are not rescaled.
```
`equlity` -> `equality`

I assume this will need to change everywhere this sentence is copied

chrisprice replied:
🤦‍♂

cprice-scottlogic committed:
refactor: don't use instanced rendering for POINTS

There's a big performance penalty to using instanced rendering
unnecessarily. This change equates the choice of drawing mode
with opting-in to instanced rendering. These are not equivalent.
However, we don't currently have a use case for distinguishing
between them.

gosthell commented:
This pull request fixes my issue #1490 and should allow me to fix my issue #1491. The code looks good to me !
